File: Day 1/exercises/data/code.R
data <- readLines("data-puzzle.txt")
numbers <- as.integer(unlist(strsplit(data[1], ",")))

boards <- data[-(1:2)]
boards <- boards[boards != ""]
board_list <- split(boards, ceiling(seq_along(boards)/5))
board_list <- lapply(board_list, function(board_lines) {
  # Split each line into numbers, handling extra spaces
  board_numbers <- unlist(lapply(board_lines, function(line) {
    as.integer(unlist(strsplit(trimws(line), "\\s+")))
  }))
  # Ensure that each board is a 5x5 matrix
  if (length(board_numbers) != 25) {
    stop("Error: Board size is not 5x5.")
  }
  matrix(board_numbers, nrow = 5, byrow = TRUE)
})

boards_filled <- lapply(board_list, function(x) matrix(0, nrow = 5, ncol = 5))

check_winner <- function(boards_filled) {
  for (i in seq_along(boards_filled)) {
    if (any(rowSums(boards_filled[[i]]) == 5) || any(colSums(boards_filled[[i]]) == 5)) {
      return(i)
    }
  }
  return(NULL)
}

for (num in numbers) {
  for (i in seq_along(board_list)) {
    boards_filled[[i]][board_list[[i]] == num] <- 1
  }
  winner <- check_winner(boards_filled)
  if (!is.null(winner)) {
    break
  }
}

if (!is.null(winner)) {
  unmarked <- board_list[[winner]][boards_filled[[winner]] == 0]
  answer <- sum(unmarked) * num
  print(glue::glue("The winner is board {winner} with a score of {answer}!"))
}

File: Day 1/exercises/data/code.py
import numpy as np
import re

def read_data():
    with open('input-data-script.txt', 'r') as f:
        data = f.read().strip().split('\n')
        numbers = [int(num) for num in re.split(',', data[0].strip())]
        boards = []
        board = []
        for i in range(2, len(data)):
            if not data[i]:
                boards.append(board)
                board = []
            else:
                board.append([int(num) for num in re.split('\\s+', data[i].strip())])
        boards=np.array(boards)
    return numbers, boards

numbers, boards = read_data()
boards_filled = np.zeros_like(boards)

def mark_number(num):
    num = int(num)
    mask = (boards == num)
    boards_filled[mask] = 1

def return_winner(boards_filled=boards_filled):
    vert = (boards_filled.sum(axis=1)==5)
    hor = (boards_filled.sum(axis=2)==5)
    if np.any(vert):
        winner = np.where(boards_filled.sum(axis=1)==5)[0]
    elif np.any(hor):
        winner = np.where(boards_filled.sum(axis=2)==5)[0]
    else:
        return None
    print("we have a winner!")
    return winner

for num in numbers:
    mark_number(num)
    vert = (boards_filled.sum(axis=1)==5)
    hor = (boards_filled.sum(axis=2)==5)
    winner = return_winner(boards_filled)
    if winner is not None:
        break

if np.any(vert):
    winner = np.where(boards_filled.sum(axis=1)==5)[0]
else:
    winner = np.where(boards_filled.sum(axis=2)==5)[0]
    
unmarked = (boards_filled[winner]==0)
answer = boards[winner][unmarked].sum()*num
print(f"The winner is board {winner+1} with a score of {answer}!")